### DISCUSSION TITLE
AVA and testing in real browsers

### DISCUSSION THREAD

**Discussion by make-github-pseudonymous-again** [avajs/ava#3022] [Ideas]
Reading [the section on browser testing](https://github.com/avajs/ava/blob/main/docs/recipes/browser-testing.md) it seems that AVA cannot be used to test code running in a real browser. [I have drafted an incomplete clone of AVA that runs tests in real browsers](https://github.com/xn-js-b9bae/xn--mxaac) via `playwright` and `rollup`. The `playwright` part replaces workers and `rollup` is used to get tests written in TypeScript to run in the browser. There are some caveats. I would love feedback.

**Comment by novemberborn** [comment]
Neat!

Is there anything we can do to make it so you don't have to clone AVA?

**Reply by make-github-pseudonymous-again** [reply]
> Is there anything we can do to make it so you don't have to clone AVA?

First thing would be to design a common interface for workers and tests' sources?

Regarding workers, it would make sense to allow concurrent tabs testing (I have yet to implement this). The AVA equivalent would be concurrent worker testing with some sort of communication possible between workers. Not sure this is in scope with AVA. If you drop that feature, then it's just a question of allowing execution environment configuration: current AVA uses whatever `node` executable is made available (maybe I am wrong and this is actually explicitly configurable?). Allowing tests in real browsers should allow to specify which browser should be used, for instance by forwarding `playwright`'s `Browser` and `Device` options.

Regarding sources, AVA v4 has dropped support for out-of-the box on-the-fly sources transpilation, because that can be configured rather generically through Node's `--experimental-loader`. Not sure things are as easy for browsers. I am currently using `rollup` to bundle tests sources and serve them via HTTP.

**Reply by novemberborn** [reply]
All the internals are still there for the source transpilation. There's no generic interface to it though, so that could be the solution.

Same for the executors, there is no generic interface but that could be done.

**Reply by make-github-pseudonymous-again** [reply]
I also have to note that I currently have no solution for code coverage. Everything became so simple for Node with `c8` but it seems that for the browser it will be back to userland code instrumentation solution... Unless there's a way to handle it easily for chromium through playwright. I do not know.
